awesome/src/common/object.rs

```diff
@@ -89,35 +116,40 @@ impl<'lua> ObjectBuilder<'lua> {
 ///
 /// You can't do anything to the object until it has been converted into a
 /// canonical form using this trait.
-pub trait Objectable<'lua, T, S: UserData + Default + Display + Send> {
-    fn cast(obj: Object<'lua>) -> rlua::Result<T> {
-        if obj.object.is::<S>()? {
-            Ok(Self::_wrap(obj))
+impl<'lua, S: State> Object<'lua, S> {
+    pub fn cast(obj: AnyUserData<'lua>) -> rlua::Result<Self> {
+        if obj.is::<S>()? {
+            Ok(obj.into())
         } else {
             use rlua::Error::RuntimeError;
             Err(RuntimeError("Could not cast object to concrete type".into()))
         }
     }
 
-    /// Given the internal object table, constructs a concrete object out
-    /// of it.
-    ///
-    /// This is only used internally, which is why it's prefixed with a "_"
-    /// Please do not use it outside of object.rs.
-    fn _wrap(table: Object<'lua>) -> T;
-
-    fn state(&self) -> rlua::Result<::std::cell::Ref<S>>;
+    pub fn state(&self) -> rlua::Result<cell::Ref<S>> {
+        Ok(self.obj.borrow::<S>()?)
+    }
 
     /// Gets a mutable reference to the internal state for the concrete object.
-    fn get_object_mut(&mut self) -> rlua::Result<::std::cell::RefMut<S>>;
+    pub fn get_object_mut(&mut self) -> rlua::Result<cell::RefMut<S>> {
+        Ok(self.obj.borrow_mut::<S>()?)
+    }
+
+    pub fn signals(&self) -> rlua::Result<rlua::Table<'lua>> {
+        self.table()?.get::<_, Table>("signals")
+    }
+
+    pub fn table(&self) -> rlua::Result<Table<'lua>> {
```

Comment:
I divided the two use cases (namely metatable and side values) into two sets of methods. See if you find this satisfying.

Reply:
Hmm, I don't really like the `{get,set}_associated_data` functions. They are still leaky because you need to still call `get_user_value` to get the actual table (which you need to do in some places, leading to that ugly call down below of `obj.obj.get_user_value()`.

You _could_ make a function to wrap that too, but what about if you want to remove an entry? No, I think the best thing is to not wrap over it, we aren't getting anything from abstracting out (leakily) that we are using the Lua user value to store these values.

...

[also you are calling get here not set](https://github.com/way-cooler/way-cooler/pull/585/files#diff-51c95dcc00f2315364dd7c72271ec263R138)

Reply:
Sorry, I can't follow. Why would accessing the table and mutating it directly be leaky (unless lua does not GC the previous value of the property, which would be strange)? Also, I don't see the benefit of calling the `get_user_value explicitly`, if it is still leaky. It makes the flaw more obvious but makes the rest of the code less expressive. This also means that if in the future someone finds a better way to store side values, more changes will be needed in the code.

Reply:
Also, for the line you pointed, I get the userdata as a table then set an index of this table. This is two calls chained, so maybe it would be clearer on separate lines.

Reply:
> Also, for the line you pointed, I get the userdata as a table then set an index of this table. This is two calls chained, so maybe it would be clearer on separate lines.

 Ah ok I see know. Thanks for clarifying.

Ok, I've been thinking about it and it seems fine. I realize I confused a use of `get_user_value` with subverting this API but it was actually different (in button there's no button object, only the class which is why I was confused). So if it properly abstracts everything now I'm probably just worrying too much about the abstraction.

Can you remove the `pub` on the `obj` in `Object` though? It's not necessary now.

Reply:
Right. Any other final touch to include in the commit?

Reply:
Nah all good. There's a few other nits in terms of style that are really small (trailing whitespace, spacing of args) but since I don't have a style guide I'm not going to bother enforcing that now.